src/main/java/com/google/sps/servlets/GmailServlet.java

```diff
@@ -66,15 +74,132 @@ public void doGet(
+   */
+  private int getUnreadEmailCountHours(int mHours, GmailClient gmailClient) throws IOException {
+    String searchQuery = GmailClient.emailQueryString(mHours, "h", true, false, "");
+    return gmailClient.listUserMessages(searchQuery).size();
+  }
+
+  /**
+   * Get the amount of unread, important emails from the last n days in a user's Gmail account
+   *
+   * @param nDays number of days to check the user's account for
+   * @param gmailClient Gmail service with valid google credential
+   * @return number of unread, important emails from last nDays days
+   * @throws IOException if an issue occurs with the Gmail Service
+   */
+  private int getImportantUnreadEmailCount(int nDays, GmailClient gmailClient) throws IOException {
+    String searchQuery = GmailClient.emailQueryString(nDays, "d", true, true, "");
+    return gmailClient.listUserMessages(searchQuery).size();
+  }
+
+  /**
+   * Get the sender of the most unread emails from the last n days in a user's Gmail account
+   *
+   * @param nDays number of days to check the user's account for
+   * @param gmailClient Gmail service with valid google credential
+   * @return sender of the most unread emails from the last nDays days
+   * @throws IOException if an issue occurs with the Gmail Service
+   */
+  private String getMostFrequentSender(int nDays, GmailClient gmailClient) throws IOException {
+    // Get user's messages - if none present, return "" as there is no sender to extract
+    String searchQuery = GmailClient.emailQueryString(nDays, "d", true, false, "");
+    GmailClient.MessageFormat messageFormat = GmailClient.MessageFormat.METADATA;
+    List<Message> unreadEmails;
+    unreadEmails =
+        gmailClient.listUserMessages(searchQuery).stream()
+            .map(
+                (Message m) -> {
+                  try {
+                    return gmailClient.getUserMessage(m.getId(), messageFormat);
+                  } catch (IOException e) {
+                    throw new RuntimeException(e);
+                  }
+                })
+            .collect(Collectors.toList());
+    if (unreadEmails.isEmpty()) {
+      return "";
+    }
+
+    // Find out which sender sent the most emails
+    // Extracted from headers within messages.
+    // Headers have the name "From" and two types of values:
+    // <sampleemail@sample.com> (if name is not available)
+    // OR
+    // Sample Sender <sampleemail@sample.com>
+    // If a name is available, this is extracted. Otherwise, the email is extracted
+    HashMap<String, Integer> senders = new HashMap<>();
+    unreadEmails.forEach(
+        (Message m) -> {
+          List<MessagePartHeader> senderList = m.getPayload().getHeaders();
+          senderList =
+              senderList.stream()
+                  .filter((MessagePartHeader header) -> header.getName().equals("From"))
+                  .collect(Collectors.toList());
+
+          String sender = senderList.get(0).getValue();
```

Comment:
[findFirst](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#findFirst--)

Reply:
Cool - thanks for this!